File: projects/sdk-go/manifest/manifest.go
// Package manifest parses orca-plugin.toml. Wire-compatible with
// projects/sdk/src/manifest.rs (the Rust reference). Every language SDK port
// must accept the canonical fixture verbatim.
package manifest

import (
	"bytes"
	"errors"
	"fmt"
	"os"
	"strconv"
	"strings"

	"github.com/pelletier/go-toml/v2"
)

// Filename is the conventional manifest filename.
const Filename = "orca-plugin.toml"

// Manifest is the parsed top-level orca-plugin.toml.
type Manifest struct {
	Plugin       PluginSection    `toml:"plugin"`
	Runtime      RuntimeSection   `toml:"runtime"`
	Surfaces     SurfacesSection  `toml:"surfaces"`
	Capabilities []CapabilityDecl `toml:"capabilities"`
}

// PluginSection — `[plugin]`.
type PluginSection struct {
	ID             string `toml:"id"`
	Version        string `toml:"version"`
	MinOrcaVersion string `toml:"min_orca_version"`
}

// RuntimeSection — `[runtime]`. Exactly one of Binary / Image must be set.
type RuntimeSection struct {
	Binary *string `toml:"binary"`
	Image  *string `toml:"image"`
	Mode   string  `toml:"mode"`
	Eager  bool    `toml:"eager"`
}

// SurfacesSection — `[surfaces]`. All fields default to false.
type SurfacesSection struct {
	MCP        bool `toml:"mcp"`
	CLI        bool `toml:"cli"`
	UI         bool `toml:"ui"`
	Docs       bool `toml:"docs"`
	Jobs       bool `toml:"jobs"`
	Storage    bool `toml:"storage"`
	Federation bool `toml:"federation"`
}

// CapabilityDecl — one entry under `[[capabilities]]`. Sensitivity must be
// "general" or "sensitive" (matches projects/sdk/src/pki.rs::Capability).
type CapabilityDecl struct {
	Name        string `toml:"name"`
	Sensitivity string `toml:"sensitivity"`
}

// ParseString parses a TOML manifest from a string.
func ParseString(s string) (*Manifest, error) {
	dec := toml.NewDecoder(bytes.NewBufferString(s))
	dec.DisallowUnknownFields()

	var m Manifest
	if err := dec.Decode(&m); err != nil {
		return nil, fmt.Errorf("parse %s: %w", Filename, err)
	}
	if m.Runtime.Mode == "" {
		m.Runtime.Mode = "process"
	}
	if err := m.validate(); err != nil {
		return nil, err
	}
	return &m, nil
}

// ParseFile parses a TOML manifest from disk.
func ParseFile(path string) (*Manifest, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read manifest at %s: %w", path, err)
	}
	return ParseString(string(data))
}

func (m *Manifest) validate() error {
	if strings.TrimSpace(m.Plugin.ID) == "" {
		return errors.New("plugin.id must not be empty")
	}
	for _, r := range m.Plugin.ID {
		if r == ' ' || r == '\t' || r == '\n' || r == '/' || r == '\\' {
			return fmt.Errorf("plugin.id %q contains invalid characters (whitespace or path separators)", m.Plugin.ID)
		}
	}
	if err := checkSemver(m.Plugin.Version, "plugin.version"); err != nil {
		return err
	}
	if err := checkSemver(m.Plugin.MinOrcaVersion, "plugin.min_orca_version"); err != nil {
		return err
	}

	switch {
	case m.Runtime.Binary != nil && m.Runtime.Image != nil:
		return errors.New("runtime.binary and runtime.image are mutually exclusive")
	case m.Runtime.Binary == nil && m.Runtime.Image == nil:
		return errors.New("runtime requires either `binary` or `image`")
	}
	if m.Runtime.Mode != "process" {
		return fmt.Errorf("runtime.mode %q: only \"process\" is supported in v0", m.Runtime.Mode)
	}

	seen := make(map[string]struct{}, len(m.Capabilities))
	for _, c := range m.Capabilities {
		if strings.TrimSpace(c.Name) == "" {
			return errors.New("capability.name must not be empty")
		}
		if c.Sensitivity != "general" && c.Sensitivity != "sensitive" {
			return fmt.Errorf("capability %q: sensitivity must be \"general\" or \"sensitive\", got %q", c.Name, c.Sensitivity)
		}
		if _, dup := seen[c.Name]; dup {
			return fmt.Errorf("duplicate capability %q", c.Name)
		}
		seen[c.Name] = struct{}{}
	}
	return nil
}

func checkSemver(v, field string) error {
	if strings.ContainsAny(v, "-+") {
		return fmt.Errorf("%s %q: pre-release/build metadata not supported in v0", field, v)
	}
	parts := strings.Split(v, ".")
	if len(parts) == 0 {
		return fmt.Errorf("%s %q: empty", field, v)
	}
	for _, p := range parts {
		if p == "" {
			return fmt.Errorf("%s %q: empty component", field, v)
		}
		if _, err := strconv.ParseUint(p, 10, 64); err != nil {
			return fmt.Errorf("%s %q: bad numeric component %q", field, v, p)
		}
	}
	return nil
}

File: projects/sdk/kotlin/src/main/kotlin/orca/sdk/Manifest.kt
// Plugin manifest — `orca-plugin.toml`. Wire-compatible with
// projects/sdk/src/manifest.rs and the other language ports.
package orca.sdk

import java.nio.file.Files
import java.nio.file.Path
import org.tomlj.Toml
import org.tomlj.TomlParseResult
import org.tomlj.TomlTable

const val MANIFEST_FILENAME: String = "orca-plugin.toml"

data class PluginSection(
    val id: String,
    val version: String,
    val minOrcaVersion: String,
)

data class RuntimeSection(
    val binary: String?,
    val image: String?,
    val mode: String,
    val eager: Boolean,
)

data class SurfacesSection(
    val mcp: Boolean = false,
    val cli: Boolean = false,
    val ui: Boolean = false,
    val docs: Boolean = false,
    val jobs: Boolean = false,
    val storage: Boolean = false,
    val federation: Boolean = false,
)

data class CapabilityDecl(val name: String, val sensitivity: String)

/**
 * `[[depends_on]]` — peer plugin this plugin needs at runtime. The host
 * enforces presence of required deps before the plugin's tools are
 * dispatchable; optional deps degrade rather than reject.
 */
data class PluginDependency(
    val id: String,
    val minVersion: String,
    val optional: Boolean = false,
) {
    /** "<id>>=<minVersion>" — the form HelloOptions expects. */
    fun formatDep(): String = if (minVersion.isEmpty()) id else "$id>=$minVersion"
}

data class Manifest(
    val plugin: PluginSection,
    val runtime: RuntimeSection,
    val surfaces: SurfacesSection,
    val capabilities: List<CapabilityDecl>,
    val dependsOn: List<PluginDependency> = emptyList(),
) {
    /** Required deps formatted for HelloOptions.pluginsRequired. */
    fun requiredDeps(): List<String> = dependsOn.filter { !it.optional }.map { it.formatDep() }

    /** Optional deps formatted for HelloOptions.pluginsOptional. */
    fun optionalDeps(): List<String> = dependsOn.filter { it.optional }.map { it.formatDep() }
}

private val ALLOWED_TOP = setOf("plugin", "runtime", "surfaces", "capabilities", "depends_on")
private val ALLOWED_PLUGIN = setOf("id", "version", "min_orca_version")
private val ALLOWED_RUNTIME = setOf("binary", "image", "mode", "eager")
private val ALLOWED_SURFACES = setOf("mcp", "cli", "ui", "docs", "jobs", "storage", "federation")
private val ALLOWED_CAPABILITY = setOf("name", "sensitivity")
private val ALLOWED_DEPEND = setOf("id", "min_version", "optional")

object ManifestParser {
    fun parseString(s: String): Manifest {
        val parsed: TomlParseResult = Toml.parse(s)
        if (parsed.hasErrors()) {
            error("parse $MANIFEST_FILENAME: ${parsed.errors().joinToString("; ")}")
        }
        denyUnknown(parsed.keySet(), ALLOWED_TOP, "top-level")

        val pluginTable = parsed.getTable("plugin") ?: error("missing [plugin]")
        denyUnknown(pluginTable.keySet(), ALLOWED_PLUGIN, "[plugin]")
        val id = (pluginTable.getString("id") ?: "").trim()
        if (id.isEmpty()) error("plugin.id must not be empty")
        for (ch in id) {
            if (ch == ' ' || ch == '\t' || ch == '\n' || ch == '/' || ch == '\\') {
                error("plugin.id \"$id\" contains invalid characters (whitespace or path separators)")
            }
        }
        val version = pluginTable.getString("version") ?: error("plugin.version missing")
        val minOrca = pluginTable.getString("min_orca_version") ?: error("plugin.min_orca_version missing")
        checkSemver(version, "plugin.version")
        checkSemver(minOrca, "plugin.min_orca_version")

        val runtimeTable = parsed.getTable("runtime") ?: error("missing [runtime]")
        denyUnknown(runtimeTable.keySet(), ALLOWED_RUNTIME, "[runtime]")
        val binary = runtimeTable.getString("binary")
        val image = runtimeTable.getString("image")
        if (binary != null && image != null) error("runtime.binary and runtime.image are mutually exclusive")
        if (binary == null && image == null) error("runtime requires either `binary` or `image`")
        val mode = runtimeTable.getString("mode") ?: "process"
        if (mode != "process") error("runtime.mode \"$mode\": only \"process\" is supported in v0")
        val eager = runtimeTable.getBoolean("eager") ?: false

        val surfacesTable = parsed.getTable("surfaces")
        if (surfacesTable != null) denyUnknown(surfacesTable.keySet(), ALLOWED_SURFACES, "[surfaces]")
        val surfaces = SurfacesSection(
            mcp = surfacesTable?.getBoolean("mcp") ?: false,
            cli = surfacesTable?.getBoolean("cli") ?: false,
            ui = surfacesTable?.getBoolean("ui") ?: false,
            docs = surfacesTable?.getBoolean("docs") ?: false,
            jobs = surfacesTable?.getBoolean("jobs") ?: false,
            storage = surfacesTable?.getBoolean("storage") ?: false,
            federation = surfacesTable?.getBoolean("federation") ?: false,
        )

        val capsArray = parsed.getArray("capabilities")
        val seen = HashSet<String>()
        val capabilities = buildList {
            if (capsArray != null) {
                for (i in 0 until capsArray.size()) {
                    val tbl = capsArray.getTable(i)
                    denyUnknown(tbl.keySet(), ALLOWED_CAPABILITY, "[[capabilities]]")
                    val name = (tbl.getString("name") ?: "").trim()
                    if (name.isEmpty()) error("capability.name must not be empty")
                    if (!seen.add(name)) error("duplicate capability \"$name\"")
                    val sensitivity = tbl.getString("sensitivity")
                        ?: error("capability \"$name\": sensitivity missing")
                    if (sensitivity != "general" && sensitivity != "sensitive") {
                        error("capability \"$name\": sensitivity must be \"general\" or \"sensitive\", got \"$sensitivity\"")
                    }
                    add(CapabilityDecl(name, sensitivity))
                }
            }
        }

        val depsArray = parsed.getArray("depends_on")
        val depIDs = HashSet<String>()
        val dependsOn = buildList {
            if (depsArray != null) {
                for (i in 0 until depsArray.size()) {
                    val tbl = depsArray.getTable(i)
                    denyUnknown(tbl.keySet(), ALLOWED_DEPEND, "[[depends_on]]")
                    val depID = (tbl.getString("id") ?: "").trim()
                    if (depID.isEmpty()) error("depends_on.id must not be empty")
                    if (depID == id) error("plugin \"$id\" cannot depend on itself")
                    if (!depIDs.add(depID)) error("duplicate dependency on \"$depID\"")
                    val minVersion = tbl.getString("min_version") ?: ""
                    checkSemver(minVersion, "depends_on[$depID].min_version")
                    val optional = tbl.getBoolean("optional") ?: false
                    add(PluginDependency(depID, minVersion, optional))
                }
            }
        }

        return Manifest(
            plugin = PluginSection(id, version, minOrca),
            runtime = RuntimeSection(binary, image, mode, eager),
            surfaces = surfaces,
            capabilities = capabilities,
            dependsOn = dependsOn,
        )
    }

    fun parseFile(path: Path): Manifest = parseString(Files.readString(path))

    private fun denyUnknown(keys: Set<String>, allowed: Set<String>, where: String) {
        for (k in keys) {
            if (k !in allowed) error("unknown field \"$k\" in $where")
        }
    }

    private fun checkSemver(v: String, field: String) {
        if (v.contains('-') || v.contains('+')) {
            error("$field \"$v\": pre-release/build metadata not supported in v0")
        }
        val parts = v.split('.')
        if (parts.isEmpty()) error("$field \"$v\": empty")
        for (p in parts) {
            if (p.isEmpty()) error("$field \"$v\": empty component")
            if (p.toULongOrNull() == null) error("$field \"$v\": bad numeric component \"$p\"")
        }
    }
}
